core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/WorkspaceBillingAccountActor.scala

```diff
@@ -0,0 +1,275 @@
+import org.broadinstitute.dsde.rawls.model._
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Implicits._
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Outcome
+import org.broadinstitute.dsde.rawls.monitor.migration.MigrationUtils.Outcome.{Failure, Success}
+
+import java.time.Instant
+import scala.concurrent.duration._
+
+object WorkspaceBillingAccountActor {
+  sealed trait WorkspaceBillingAccountsMessage
+
+  case object UpdateBillingAccounts extends WorkspaceBillingAccountsMessage
+
+  def apply(dataSource: SlickDataSource,
+            gcsDAO: GoogleServicesDAO,
+            initialDelay: FiniteDuration,
+            pollInterval: FiniteDuration): Behavior[WorkspaceBillingAccountsMessage] =
+    Behaviors.setup { context =>
+      val actor = WorkspaceBillingAccountActor(dataSource, gcsDAO)
+      Behaviors.withTimers { scheduler =>
+        scheduler.startTimerAtFixedRate(UpdateBillingAccounts, initialDelay, pollInterval)
+        Behaviors.receiveMessage {
+          case UpdateBillingAccounts =>
+            try actor.updateBillingAccounts.unsafeRunSync()
+            catch {
+              case t: Throwable => context.executionContext.reportFailure(t)
+            }
+            Behaviors.same
+        }
+      }
+    }
+}
+
+final case class WorkspaceBillingAccountActor(dataSource: SlickDataSource, gcsDAO: GoogleServicesDAO)
+  extends LazyLogging {
+
+  import dataSource.dataAccess._
+  import dataSource.dataAccess.driver.api._
+
+  /* Sync billing account changes to billing projects and their associated workspaces with Google
+   * one at a time.
+   *
+   * Why not all at once?
+   * To reduce the number of stale billing account changes we make.
+   *
+   * Let's imagine that there are lots of billing account changes to sync. If a user changes
+   * the billing account on a billing project that's currently being sync'ed then we'll have to
+   * sync the billing projects and all workspaces again immediately. If we sync one at a time
+   * we'll have a better chance of doing this once.
+   */
+  def updateBillingAccounts: IO[Unit] =
+    getABillingProjectChange.flatMap(_.traverse_ { billingAccountChange =>
+      for {
+        _ <- info("Updating Billing Account on Billing Project in Google",
+          "changeId" -> billingAccountChange.id,
+          "billingProject" -> billingAccountChange.billingProjectName.value,
+          "newBillingAccount" -> billingAccountChange.newBillingAccount.map(_.value)
+        )
+
+        billingProject <- loadBillingProject(billingAccountChange.billingProjectName)
+        billingProbeHasAccess <- billingProject.billingAccount.traverse(gcsDAO.testDMBillingAccountAccess(_).io)
+        billingProjectSyncAttempt <- syncBillingProjectWithGoogle(billingProject).attempt
+        syncOutcome = Outcome.fromEither(billingProjectSyncAttempt)
+        _ <- recordBillingProjectSyncOutcome(billingAccountChange, billingProbeHasAccess, syncOutcome)
```

Comment:
Should we record the outcome only after we've synced all the workspaces? (Since presumably it could still fail at the workspace stage.)

Reply:
Thanks - this makes sense I think, didn't realize we captured errors at the workspace/googleProject level as well.